### ISSUE TITLE
exception in ProcessDLibs

### ISSUE THREAD

**Issue by SingingBush** [intellij-dlanguage/intellij-dlanguage#609]
no detail given about what the user was doing at the time but perhaps worth seeing if this happens a lot. 

**Desktop (please provide environment info):**
 - OS: Linux (please include distro and version)]
 - Java 11.0.10 (Jetbrains)
 - IDE: Intellij IDEA CE 2020.1 beta
 - Plugin Version 1.26

```
java.lang.Throwable: OrderEntry LibraryDependency(library=LibraryId(name=objective-gl-0.0.2, tableId=com.intellij.workspaceModel.storage.bridgeEntities.LibraryTableId$ProjectLibraryTableId@c37ff29), exported=false, scope=COMPILE) does not belong to modifiableRootModel of module dlangtesting
    at com.intellij.openapi.diagnostic.Logger.error(Logger.java:161)
    at com.intellij.workspaceModel.ide.impl.legacyBridge.module.roots.ModifiableRootModelBridgeImpl.removeOrderEntry(ModifiableRootModelBridgeImpl.kt:367)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion$processDLibsImpl$1.consume(ProcessDLibs.kt:111)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion$processDLibsImpl$1.consume(ProcessDLibs.kt:75)
    at com.intellij.openapi.roots.ModuleRootModificationUtil.lambda$updateModel$8(ModuleRootModificationUtil.java:128)
    at com.intellij.openapi.roots.ModuleRootModificationUtil.modifyModel(ModuleRootModificationUtil.java:136)
    at com.intellij.openapi.roots.ModuleRootModificationUtil.updateModel(ModuleRootModificationUtil.java:127)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion.processDLibsImpl(ProcessDLibs.kt:104)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion.access$processDLibsImpl(ProcessDLibs.kt:75)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion$processDLibs$1$task$1.run(ProcessDLibs.kt:91)
    at com.intellij.openapi.progress.impl.CoreProgressManager$TaskRunnable.run(CoreProgressManager.java:998)
    at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressAsync$5(CoreProgressManager.java:497)
    at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$3(ProgressRunner.java:228)
    at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:178)
    at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:688)
    at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:634)
    at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:64)
    at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:165)
    at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:228)
    at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    at java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:668)
    at java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:665)
    at java.security.AccessController.doPrivileged(AccessController.java)
    at java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:665)
    at java.lang.Thread.run(Thread.java:834)
```

**Comment by SingingBush** [comment]
been a few of these now. The problem is here:

```kotlin
            // remove all existing libs
            ModuleRootModificationUtil.updateModel(module) { model ->
                model.orderEntries
                        .filterIsInstance<LibraryOrderEntry>()
                        .forEach {
                            it.library?.let {
                                model.moduleLibraryTable.removeLibrary(it)
                            }
                            model.removeOrderEntry(it)    <----- THIS LINE
                        }
            }
```

**Comment by SingingBush** [comment]
again in Intellij 2020.3.3 and plugin v1.26:

```
java.lang.Throwable: OrderEntry LibraryDependency(library=LibraryId(name=dpq-0.11.4+commit.2.geb2a79f, tableId=com.intellij.workspaceModel.storage.bridgeEntities.LibraryTableId$ProjectLibraryTableId@35427aa3), exported=false, scope=COMPILE) does not belong to modifiableRootModel of module PROJECTNAME
    at com.intellij.openapi.diagnostic.Logger.error(Logger.java:165)
    at com.intellij.workspaceModel.ide.impl.legacyBridge.module.roots.ModifiableRootModelBridgeImpl.removeOrderEntry(ModifiableRootModelBridgeImpl.kt:367)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion$processDLibsImpl$1.consume(ProcessDLibs.kt:111)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion$processDLibsImpl$1.consume(ProcessDLibs.kt:75)
    at com.intellij.openapi.roots.ModuleRootModificationUtil.lambda$updateModel$8(ModuleRootModificationUtil.java:126)
    at com.intellij.openapi.roots.ModuleRootModificationUtil.modifyModel(ModuleRootModificationUtil.java:134)
    at com.intellij.openapi.roots.ModuleRootModificationUtil.updateModel(ModuleRootModificationUtil.java:125)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion.processDLibsImpl(ProcessDLibs.kt:104)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion.access$processDLibsImpl(ProcessDLibs.kt:75)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion$processDLibs$1$task$1.run(ProcessDLibs.kt:91)
    at com.intellij.openapi.progress.impl.CoreProgressManager$TaskRunnable.run(CoreProgressManager.java:962)
    at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressAsync$5(CoreProgressManager.java:472)
    at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$3(ProgressRunner.java:235)
    at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:178)
    at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:658)
    at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:610)
    at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:65)
    at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:165)
    at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:235)
    at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    at java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:668)
    at java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:665)
    at java.security.AccessController.doPrivileged(AccessController.java)
    at java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:665)
    at java.lang.Thread.run(Thread.java:834)
```

**Comment by belgrid** [comment]
I always get these error messages when I open the IDE and a project is loaded. Probably after indexing the files. Thanks for taking care of this annoying bug. Is there already a release date for the next version of the plugin? There are apparently some bugs that you have already fixed but have not yet landed in the Jetbrains repo.

OS: Linux (Archlinux)]
Java 11.0.10 (Jetbrains)
IDE: Intellij IDEA CE 2021.1 CE
Plugin Version 1.26

**Comment by SingingBush** [comment]
This has not been fixed yet. It's one of two (currently) [high priority](https://github.com/intellij-dlanguage/intellij-dlanguage/labels/top%20priority) issues. I don't have availability to work on this currently so unless a pull request is made a fix is not likely to be released any time soon.

**Comment by SingingBush** [comment]
This just happened to me whilst working on another issue and in turn caused a whole bunch of exceptions in IDEA core:

```
com.intellij.workspaceModel.storage.impl.exceptions.AddDiffException: addDiff: persistent id already exists. Removing old entity
Persistent id: LibraryId(name=dunit-1.0.16, tableId=com.intellij.workspaceModel.storage.bridgeEntities.LibraryTableId$ProjectLibraryTableId@20956b7d)

Existing entity data: LibraryEntity-:-34
New entity data: LibraryEntityData(tableId=com.intellij.workspaceModel.storage.bridgeEntities.LibraryTableId$ProjectLibraryTableId@20956b7d, name=dunit-1.0.16, roots=[LibraryRoot(url=file:///home/user/.dub/packages/dunit-1.0.16/dunit/source, type=LibraryRootTypeId(name=CLASSES), inclusionOptions=ROOT_ITSELF)], excludedRoots=[], entitySource=FileInDirectory(directory=file:///home/user/projects/dfract/.idea/libraries, projectLocation=DirectoryBased(projectDir=file:///home/user/projects/dfract)), id=32, id=32)
	at com.intellij.workspaceModel.storage.impl.WorkspaceEntityStorageBuilderImpl.addDiffAndReport$intellij_platform_workspaceModel_storage(WorkspaceEntityStorageImpl.kt:558)
	at com.intellij.workspaceModel.storage.impl.AddDiffOperation.checkPersistentId(AddDiffOperation.kt:355)
	at com.intellij.workspaceModel.storage.impl.AddDiffOperation.addDiff(AddDiffOperation.kt:28)

```

The _ProcessDLibs_ stack trace:

```
java.lang.Throwable: OrderEntry LibraryDependency(library=LibraryId(name=dunit-1.0.16, tableId=com.intellij.workspaceModel.storage.bridgeEntities.LibraryTableId$ProjectLibraryTableId@20956b7d), exported=false, scope=COMPILE) does not belong to modifiableRootModel of module dfract
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:159)
	at com.intellij.workspaceModel.ide.impl.legacyBridge.module.roots.ModifiableRootModelBridgeImpl.removeOrderEntry(ModifiableRootModelBridgeImpl.kt:263)
	at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion$processDLibsImpl$1.consume(ProcessDLibs.kt:111)
	at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion$processDLibsImpl$1.consume(ProcessDLibs.kt:75)
	at com.intellij.openapi.roots.ModuleRootModificationUtil.lambda$updateModel$8(ModuleRootModificationUtil.java:126)
	at com.intellij.openapi.roots.ModuleRootModificationUtil.modifyModel(ModuleRootModificationUtil.java:134)
	at com.intellij.openapi.roots.ModuleRootModificationUtil.updateModel(ModuleRootModificationUtil.java:125)
	at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion.processDLibsImpl(ProcessDLibs.kt:104)
	at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion.access$processDLibsImpl(ProcessDLibs.kt:75)
	at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion$processDLibs$1$task$1.run(ProcessDLibs.kt:91)
	at com.intellij.openapi.progress.impl.CoreProgressManager$TaskRunnable.run(CoreProgressManager.java:962)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressAsync$5(CoreProgressManager.java:472)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$3(ProgressRunner.java:235)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:178)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:658)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:610)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:65)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:165)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:235)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
```

**Comment by SingingBush** [comment]
spent 2 hours trying to fix this. It's not a simple problem to solve. It's not particularly clear why it's started to be an issue. The bits of code that are causing problems haven't changed in over two years.

**Comment by AlexPl292** [comment]
Hi there! Is this thing reproducable with the newer build of IJ (2021.1)? I see some similar reports that are fixed in this version.
https://youtrack.jetbrains.com/issue/IDEA-267246
https://youtrack.jetbrains.com/issue/IDEA-256694

**Comment by SingingBush** [comment]
> Hi there! Is this thing reproducable with the newer build of IJ (2021.1)? I see some similar reports that are fixed in this version.
> https://youtrack.jetbrains.com/issue/IDEA-267246
> https://youtrack.jetbrains.com/issue/IDEA-256694

Yes, it's happening with the latest release as well

**Comment by SingingBush** [comment]
This is still happening in 1.26.2

```
java.lang.Throwable: OrderEntry LibraryDependency(library=LibraryId(name=colourfulmoon-1.0.2, tableId=com.intellij.workspaceModel.storage.bridgeEntities.LibraryTableId$ProjectLibraryTableId@17d67e96), exported=false, scope=COMPILE) does not belong to modifiableRootModel of module eloquent
    at com.intellij.openapi.diagnostic.Logger.error(Logger.java:182)
    at com.intellij.workspaceModel.ide.impl.legacyBridge.module.roots.ModifiableRootModelBridgeImpl.removeOrderEntry(ModifiableRootModelBridgeImpl.kt:388)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion$processDLibsImpl$1.consume(ProcessDLibs.kt:111)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion$processDLibsImpl$1.consume(ProcessDLibs.kt:75)
    at com.intellij.openapi.roots.ModuleRootModificationUtil.lambda$updateModel$9(ModuleRootModificationUtil.java:135)
    at com.intellij.openapi.roots.ModuleRootModificationUtil.modifyModel(ModuleRootModificationUtil.java:143)
    at com.intellij.openapi.roots.ModuleRootModificationUtil.updateModel(ModuleRootModificationUtil.java:134)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion.processDLibsImpl(ProcessDLibs.kt:104)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion.access$processDLibsImpl(ProcessDLibs.kt:75)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion$processDLibs$1$task$1.run(ProcessDLibs.kt:91)
```

**Comment by SingingBush** [comment]
reported from v1.27, Intellij 2021.3 on Linux (December 15th):

```
java.lang.Throwable: OrderEntry LibraryDependency(library=LibraryId(name=phobos, tableId=com.intellij.workspaceModel.storage.bridgeEntities.LibraryTableId$ProjectLibraryTableId@c6750aa), exported=false, scope=COMPILE) does not belong to modifiableRootModel of module helloD
    at com.intellij.openapi.diagnostic.Logger.error(Logger.java:182)
    at com.intellij.workspaceModel.ide.impl.legacyBridge.module.roots.ModifiableRootModelBridgeImpl.removeOrderEntry(ModifiableRootModelBridgeImpl.kt:397)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion.processDLibsImpl$lambda-3(ProcessDLibs.kt:111)
    at com.intellij.openapi.roots.ModuleRootModificationUtil.lambda$updateModel$9(ModuleRootModificationUtil.java:135)
    at com.intellij.openapi.roots.ModuleRootModificationUtil.modifyModel(ModuleRootModificationUtil.java:143)
    at com.intellij.openapi.roots.ModuleRootModificationUtil.updateModel(ModuleRootModificationUtil.java:134)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion.processDLibsImpl(ProcessDLibs.kt:104)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion.access$processDLibsImpl(ProcessDLibs.kt:75)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion$processDLibs$1$task$1.run(ProcessDLibs.kt:91)
    at com.intellij.openapi.progress.impl.CoreProgressManager.startTask(CoreProgressManager.java:436)
    at com.intellij.openapi.progress.impl.ProgressManagerImpl.startTask(ProgressManagerImpl.java:120)
    at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressAsync$5(CoreProgressManager.java:496)
    at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$3(ProgressRunner.java:244)
    at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:188)
    at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$12(CoreProgressManager.java:624)
    at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:698)
    at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:646)
    at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:623)
    at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:66)
    at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:175)
    at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:244)
```

**Comment by SingingBush** [comment]
plugin version 1.27 on Intellij 2021.2:

```
java.lang.Throwable: OrderEntry LibraryDependency(library=LibraryId(name=raylib-d-4.0.0, tableId=com.intellij.workspaceModel.storage.bridgeEntities.LibraryTableId$ProjectLibraryTableId@7efdaf87), exported=false, scope=COMPILE) does not belong to modifiableRootModel of module ******************
    at com.intellij.openapi.diagnostic.Logger.error(Logger.java:182)
    at com.intellij.workspaceModel.ide.impl.legacyBridge.module.roots.ModifiableRootModelBridgeImpl.removeOrderEntry(ModifiableRootModelBridgeImpl.kt:364)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion.processDLibsImpl$lambda-3(ProcessDLibs.kt:111)
    at com.intellij.openapi.roots.ModuleRootModificationUtil.lambda$updateModel$9(ModuleRootModificationUtil.java:135)
    at com.intellij.openapi.roots.ModuleRootModificationUtil.modifyModel(ModuleRootModificationUtil.java:143)
    at com.intellij.openapi.roots.ModuleRootModificationUtil.updateModel(ModuleRootModificationUtil.java:134)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion.processDLibsImpl(ProcessDLibs.kt:104)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion.access$processDLibsImpl(ProcessDLibs.kt:75)
    at io.github.intellij.dlanguage.actions.ProcessDLibs$Companion$processDLibs$1$task$1.run(ProcessDLibs.kt:91)
```

**Comment by AlexPl292** [comment]
Hi, I've checked the code and I'm not sure why it worked previously without the exception. The module library is removed from the library table if it's removed from the order entries and vice versa. So, in the plugin you can just leave 
```
.forEach {
    model.removeOrderEntry(it)
}
```

**Comment by SingingBush** [comment]
I tried that previously and ended up getting errors elsewhere. Further down in the same class there is a _createLibraryDependency_ method with the line `val library = projectLibraryModel.createLibrary(libraryName)` which will throw PersistentIdAlreadyExistsException. I tried preventing that by preceeding it with

```java
            projectLibraryModel.getLibraryByName(libraryName)?.let {
                projectLibraryModel.removeLibrary(it) 
            }
```

or even using a unique name by calling `LibraryEditingUtil.suggestNewLibraryName(projectLibraryModel, libraryName)` but couldn't get the right outcome. 

Essentially I want to be able to import a dub project and have the dependencies be detected as the project's libraries. Making subsequent calls to the ProcessDLibs action (available from tools menu)
![image](https://user-images.githubusercontent.com/870567/150436511-cf1d44b2-2a6a-4755-87f9-5a6509ba3f4e.png)
should not end up duplicating libraries. There was previously a problem in which libraries were getting duplicated in the project structure:
![image](https://user-images.githubusercontent.com/870567/150436814-a4c4d743-a3d3-4ae2-a7b0-41630e8a5372.png)

**Comment by chashnikov** [comment]
@SingingBush `ModifiableRootModel::removeOrderEntry` removes dependency from the module to a library. If the library is defined at the module level, it also removes the library itself. `projectLibraryModel.createLibrary` creates a project-level library, in that case `ModifiableRootModel::removeOrderEntry` will remove the dependency but won't remove the library (because it may be added as a dependency in other modules). If you want to remove the library as well, you need to do it manually via `projectLibraryTable`.

BTW I see a problem with the current implementation of [`createLibraryDependency`](https://github.com/intellij-dlanguage/intellij-dlanguage/blob/c94a4f5463c892f8919cfb1e71f168c8d3c100ac/src/main/kotlin/io/github/intellij/dlanguage/actions/ProcessDLibs.kt#L189): it always creates modifiable models, but commits them only if `sources != null`. This may cause memory leaks, if the models aren't committed they must be disposed, see javadoc of [`LibraryTable::getModifiableModel`](https://github.com/JetBrains/intellij-community/blob/056ac17679376c66b840072948a044fdb114f84c/platform/projectModel-api/src/com/intellij/openapi/roots/libraries/LibraryTable.java#L47).

**Comment by SingingBush** [comment]
@chashnikov thanks for taking a look. So from what you're saying I should be removing from the project using something like:

```
// remove library at the project level:
val libraryTable: LibraryTable? = LibraryTablesRegistrar.getInstance()
    .getLibraryTableByLevel(LibraryTablesRegistrar.PROJECT_LEVEL, project)

libraryTable?.modifiableModel?.libraries?.iterator()?
    .forEach { libraryTable.removeLibrary(it) }
```

**Comment by SingingBush** [comment]
@AlexPl292 this appears to be resolved now so I'm considering releasing 1.27.1 soon. If you have time please pull the latest changes in _develop_ branch and try running/testing locally.
